src/include/ceph_fs.h

```diff
@@ -629,7 +629,8 @@ struct ceph_mds_request_head {
 	__le64 oldest_client_tid;
 	__le32 mdsmap_epoch;           /* on client */
 	__le32 flags;                  /* CEPH_MDS_FLAG_* */
-	__u8 num_retry, num_fwd;       /* count retry, fwd attempts */
+	__u32 num_retry;               /* count retry */
+	__u32 num_fwd;                 /* fwd attempts */
```

Comment:
If these types don't match somewhere else that they're used, that seems like a big problem.

But we can't change these types in this struct like this, because they're part of the wire format! See the preceding comment: "Note that any change to this structure must ensure that it is compatible with ceph_mds_request_head_legacy". Making changes to this part of the wire protocol needs to be done with extreme caution and a lot of ancillary changes to handle backwards compatibility.

Reply:
Yeah, right. 

The issue here is that when the MDS1 needs to forward the request it will transfer the `num_fwd + 1` in `MClientRequestForward` using `uint32_t` to client and the client will retry the request to MDS2, but this time it will transfer the `num_fwd` in `ceph_mds_request_head` by using `__u8`.

So during the looping the number will be truncated and the client will go into insane mode in `handle_forward()` in kernel, for example, in Line 3311:

```
3285 static void handle_forward(struct ceph_mds_client *mdsc,
3286                            struct ceph_mds_session *session,                                                                                      
3287                            struct ceph_msg *msg)
3288 {
 ...
3308         if (test_bit(CEPH_MDS_R_ABORTED, &req->r_req_flags)) {
3309                 dout("forward tid %llu aborted, unregistering\n", tid);
3310                 __unregister_request(mdsc, req);
3311         } else if (fwd_seq <= req->r_num_fwd) {
3312                 dout("forward tid %llu to mds%d - old seq %d <= %d\n",
3313                      tid, next_mds, req->r_num_fwd, fwd_seq);
3314         } else {
3315                 /* resend. forward race not possible; mds would drop */
3316                 dout("forward tid %llu to mds%d (we resend)\n", tid, next_mds);
3317                 BUG_ON(req->r_err);
3318                 BUG_ON(test_bit(CEPH_MDS_R_GOT_RESULT, &req->r_req_flags));
3319                 req->r_attempts = 0;
3320                 req->r_num_fwd = fwd_seq;
3321                 req->r_resend_mds = next_mds;
3322                 put_request_session(req);
3323                 __do_request(mdsc, req);
3324         }
3325         ceph_mdsc_put_request(req);
...
}
```

It just assumes it is a old request and ignore it, but the kclient will wait for the request to finish *forever*.

Reply:
There has 2 methods to fix this:

1, just likes in this PR to switch them to `__u32` instead in `ceph_mds_request_head`.

2, limit the max forward number to 256.

Reply:
This issue not only will happen in https://github.com/ceph/ceph/pull/45073. When the metadata pool reaches full ratio we could hit this issue too in `void Server::dispatch_client_request(MDRequestRef& mdr) {}`.

Reply:
can't we stop forwarding the request if the client ever receives `num_fwd` in `ceph_mds_request_head` as `0` ?
in other words, is it okay to presume that when client receives `num_fwd` in `ceph_mds_request_head` as zero, then the request has already hopped around 256 times ? i.e. there has been a wrap around of the truncated 256 bit value

Reply:
I am not very sure, from the currently client side code it seems possible that old reply with smaller seq could come after the seq has increased. So we need to make sure the seq recorded in `request` in client side equals to 256 and the new coming seq is smaller than 256.

As a workaround currently in https://github.com/ceph/ceph/pull/45728 and https://github.com/ceph/ceph/pull/45688 I have hard limited the retry times to 256. This workaround always makes sense for old ceph even we fix this here.

Reply:
ceph_mds_request_head is an on-the-wire protocol structure. Also, these particular fields are part of ceph_mds_request_head_legacy, which was unversioned, so you can't just change them like this without breaking backward compatibility.

If you want to do this, then I think you'll have to bump the version field in ceph_mds_request_head and add new __le32 fields to the end of the struct for these values. Then clients and servers that understand the new format can just ignore the old values in lieu of these.

Reply:
Yeah, right.  I am still thinking whether should we add one new feature bit for this.